rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+
+## How do I manage package dependencies? {#rse-r-package-dependencies}
+
+Let's add another function in another file.
+(We don't really need the function,
+and it doesn't have to go in another file,
+but we'll create it for tutorial purposes.)
+
+```{r, child="rse-r-package/etc/counter-original.R"}
+```
+
+This function depends on functions from the tidyverse---more specifically,
+from the dplyr package.
+Since our compiled-and-distributable package will only contain the compiled code for its own functions,
+direct calls to functions from other libraries won't work after the package is installed.
+To fix this,
+we add this to the roxygen2 comment block for our function:
+
+```{r eval=FALSE}
+#' @import dplyr
+#' @importFrom magrittr %>%
+```
+
+to tell the package builder that we want all of dplyr available,
+plus the `%>%` operator from magrittr.
+We then various calls to use their package prefix:
+
+```{r eval=FALSE}
+tally_words <- function(all_words) {
+  dplyr::tibble(word = all_words) %>%
+    dplyr::group_by(word) %>%
+    dplyr::tally()
+}
+```
+
+Running a check gives us the following error:
+
+```
+❯ checking R code for possible problems ... NOTE
+  tally_words: no visible global function definition for ‘%>%’
+  tally_words: no visible global function definition for ‘tibble’
+  tally_words: no visible global function definition for ‘group_by’
+  tally_words: no visible binding for global variable ‘word’
+  tally_words: no visible global function definition for ‘tally’
+  Undefined global functions or variables:
+    %>% group_by tally tibble word
+```
+
+What we need to do is add these three lines to the bottom of `DESCRIPTION`:
+
+```text
+Imports:
+    dplyr (>= 0.7.0),
+    magrittr (>= 1.5.0)
+```
+
+The `Imports` field in `DESCRIPTION` actually has nothing to do with importing functions;
+it just ensures that those packages are installed when this package is.
+As for the version numbers in parentheses,
+we got those by running `packageVersion("readr")` and similar commands inside RStudio
+and then rounding off.
+(See Section \@ref(branches-tag) for a discussion of version numbering.)
+
+All right: are we done now?
+No, we are not:
+
+```text
+❯ checking R code for possible problems ... NOTE
+  tally_words: no visible binding for global variable ‘word’
+  Undefined global functions or variables:
+    word
+```
+
+This is annoying but understandable.
+When the package builder is checking our code,
+it has no idea what columns are going to be in our data frames,
+so it has no way to know if `group_by(word)` will cause a problem.
+However,
+this is just a `NOTE`, not an `ERROR`,
+so we can try running "Build...Install and Restart"
+to build our package,
+re-start our R session (so that memory is clean),
+and load our newly-created package.
+It works!
+
+But there's still that warning.
+To fix it,
+we add this to our roxygen2 comment block:
+
+```{r eval=FALSE}
+#' @importFrom rlang .data
```

Comment:
And also: `usethis::use_package("rlang")` to add to DESCRIPTION

Reply:
As above.